contrib/babelfishpg_tsql/src/rolecmds.c

```diff
@@ -1733,3 +1755,154 @@ has_user_in_db(const char *login, char **db_name)
+{
+	char *pos_slash = NULL;
+
+	if ((pos_slash = strchr(input, '\\')) != NULL)
+	{
+		/*
+		 * This means that provided login name is in windows format 
+		 * so let's update role_name with UPN format.
+		 */
+		return psprintf("%s@%s", 
+						str_tolower(pos_slash + 1, strlen(pos_slash + 1), C_COLLATION_OID),
+						str_toupper(input, (pos_slash - input), C_COLLATION_OID));
+	}
+	else
+		return input;
+}
+
+/*
+ * get_roleform_ext - Useful when someone tries to drop login and that login is in
+ * the form of windows format i.e, domain\login
+ */
+HeapTuple 
+get_roleform_ext(char *login)
+{
+	Relation	bbf_authid_login_ext_rel;
+	HeapTuple	tuple;
+	ScanKeyData	scanKey;
+	SysScanDesc	scan;
+	char		*upn_login;
+	TupleDesc	dsc;
+
+	/* 
+	 * If login being dropped is not valid windows format
+	 * then return the invalid tuple from here only.
+	 */
+	if (!strchr(login, '\\'))
+		return NULL;
+
+	upn_login = convertToUPN(login);
+
+	/*
+	 * Trying to lookup provided windows user in babelfish_authid_login_ext catalog
+	 * using UPN form.
+	 */
+
+	/* Fetch the relation sys.babelfish_authid_login_ext */
+	bbf_authid_login_ext_rel = table_open(get_authid_login_ext_oid(),
+										  RowExclusiveLock);
+	dsc = RelationGetDescr(bbf_authid_login_ext_rel);
+
+	/* Search and drop on the role */
+	ScanKeyInit(&scanKey,
+				Anum_bbf_authid_login_ext_rolname,
+				BTEqualStrategyNumber, F_NAMEEQ,
+				CStringGetDatum(upn_login));
+
+	scan = systable_beginscan(bbf_authid_login_ext_rel,
+							  get_authid_login_ext_idx_oid(),
+							  true, NULL, 1, &scanKey);
+
+	tuple = systable_getnext(scan);
+
+	if (HeapTupleIsValid(tuple))
+	{
+		char *type;
+		bool isnull = true;
+		Datum datum = heap_getattr(tuple, Anum_bbf_authid_login_ext_type, dsc, &isnull);
+		if (isnull)
+			tuple = NULL;
+		type = pstrdup(TextDatumGetCString(datum));
+		/* Not a windows login */
+		if (strcasecmp(type, "U") != 0)
+			tuple = NULL;
+		pfree(type);
+	}
+
+	systable_endscan(scan);
+	table_close(bbf_authid_login_ext_rel, AccessShareLock);
+
+	if (!HeapTupleIsValid(tuple))
+	{
+		if (upn_login != login)
+			pfree(upn_login);
+		return tuple;
+	}
+
+	/* It is proven that this rolename is indeed windows login. */
+	tuple = SearchSysCache1(AUTHNAME, PointerGetDatum(upn_login));
+
+	if (upn_login != login)
+		pfree(upn_login);
+
+	/* Return tuple even if it is invalid tuple. */
+	return tuple;
+}
+
+/*
+* AD does not allow user to have some special characters,
+* from babelfish side, we can not connect to AD directly 
+* so we do not know whether the user exists in AD or not.
+* From TDS endpoint, if login is created with such special
+* characters, then we will throw error or else the user will 
+* get confused because the login will get created but they 
+* won't be able to connect
+*/
+bool 
+windows_login_contains_invalid_chars(char* input)
+{
+	char* pos_slash = strchr(input, '\\');
+	
+	char* logon_name = pos_slash + 1;
+
+	int start = 0;
+	while(logon_name[start] != '\0'){
```

Comment:
space after while and bracket should start from next line. And rename variable `start` with something like `iter` or simple `i`

Reply:
Done!